pkg/kubelet/volume/iscsi_pd/iscsi_pd.go

```diff
@@ -0,0 +1,189 @@
+	"github.com/golang/glog"
+)
+
+// This is the primary entrypoint for volume plugins.
+func ProbeVolumePlugins() []volume.Plugin {
+	return []volume.Plugin{&ISCSIDiskPlugin{nil, false, "", "", ""},
+		&ISCSIDiskPlugin{nil, true, "", "", ""}}
+}
+
+type ISCSIDiskPlugin struct {
+	host       volume.Host
+	legacyMode bool // if set, plugin answers to the legacy name
+	portal     string
+	iqn        string
+	lun        string
+}
+
+var _ volume.Plugin = &ISCSIDiskPlugin{}
+
+const (
+	ISCSIDiskPluginName       = "kubernetes.io/iscsi-pd"
+	ISCSIDiskPluginLegacyName = "iscsi-pd"
+)
+
+func (plugin *ISCSIDiskPlugin) Init(host volume.Host) {
+	plugin.host = host
+}
+
+func (plugin *ISCSIDiskPlugin) Name() string {
+	if plugin.legacyMode {
+		return ISCSIDiskPluginLegacyName
+	}
+	return ISCSIDiskPluginName
+}
+
+func (plugin *ISCSIDiskPlugin) CanSupport(spec *api.Volume) bool {
+	if plugin.legacyMode {
+		// Legacy mode instances can be cleaned up but not created anew.
+		return false
+	}
+
+	if spec.ISCSIDisk != nil {
+		return true
+	}
+	return false
+}
+
+func (plugin *ISCSIDiskPlugin) NewBuilder(spec *api.Volume, podUID types.UID) (volume.Builder, error) {
+	// Inject real implementations here, test through the internal function.
+	return plugin.newBuilderInternal(spec, podUID, &ISCSIDiskUtil{}, mount.New(), exec.New())
+}
+
+func (plugin *ISCSIDiskPlugin) newBuilderInternal(spec *api.Volume, podUID types.UID, manager disk.PDManager, mounter mount.Interface, exe exec.Interface) (volume.Builder, error) {
+	if plugin.legacyMode {
+		// Legacy mode instances can be cleaned up but not created anew.
+		return nil, fmt.Errorf("legacy mode: can not create new instances")
+	}
+	lun := "0"
+	if spec.ISCSIDisk.Lun != 0 {
+		lun = strconv.Itoa(spec.ISCSIDisk.Lun)
+	}
+	plugin.portal = spec.ISCSIDisk.Portal
+	plugin.iqn = spec.ISCSIDisk.IQN
+	plugin.lun = lun
+	return &iscsiDisk{
+		podUID:     podUID,
+		volName:    spec.Name,
+		portal:     spec.ISCSIDisk.Portal,
+		iqn:        spec.ISCSIDisk.IQN,
+		lun:        lun,
+		fsType:     spec.ISCSIDisk.FSType,
+		readOnly:   spec.ISCSIDisk.ReadOnly,
+		exec:       exe,
+		manager:    manager,
+		mounter:    mounter,
+		legacyMode: false,
+		plugin:     plugin,
+	}, nil
+}
+
+func (plugin *ISCSIDiskPlugin) NewCleaner(volName string, podUID types.UID) (volume.Cleaner, error) {
+	// Inject real implementations here, test through the internal function.
+	return plugin.newCleanerInternal(volName, podUID, &ISCSIDiskUtil{}, mount.New(), exec.New())
+}
+
+func (plugin *ISCSIDiskPlugin) newCleanerInternal(volName string, podUID types.UID, manager disk.PDManager, mounter mount.Interface, exe exec.Interface) (volume.Cleaner, error) {
+	legacy := false
+	if plugin.legacyMode {
+		legacy = true
+	}
+
+	return &iscsiDisk{
+		podUID:     podUID,
+		volName:    volName,
+		legacyMode: legacy,
+		manager:    manager,
+		mounter:    mounter,
+		plugin:     plugin,
+		portal:     plugin.portal,
```

Comment:
Ahh, I see what you're doing.  NAK.  Cleaner has to handle the case of kubelet crashing, and when it recovers it finds a volume directory that it has no Pod for.  any state you need in Cleaner MUST be recoverable from the system.  Either use the name or dir to look up the mount info in /proc/mounts or else embed info into the name you write to disk.

For example, given an iSCSI with IQN of "x" and LUN of "y" you might make the global mountpoint be iscsi/x_y - then you could parse out IQN and LUN later.

But do you really need these fields for teardown?